src/foolscap/slicers/root.py

```diff
@@ -54,22 +68,56 @@ def slicerForObject(self, obj):
             s = tokens.ISlicer(copier)
             return s
 
-        slicerFactory = self.slicerTable.get(type(obj))
+        slicerFactory = None
+        
+        # PY3KPORT
+        if six.PY2:
+            slicerFactory = self.slicerTable.get(type(obj))
+        elif six.PY3:
+            if type(obj) not in self.slicerTable:
+                # Determine if this is an instance
+                # Unfortunately this is True for file objects as well
+                # so we need another check as file objects have a different slicer
+                if hasattr(obj, '__class__') and isinstance(obj.__class__, type) and (not hasattr(obj, 'fileno')):
+                    # This is a class instance
+                    slicerFactory = self.slicerTable.get(InstanceType)
+
+            # Also for class methods i.e if obj == C.f where C is a
+            # class and f is its method, in Py3 the type evaluates
+            # to 'function' not 'method' whereas in Py2 it returns
+            # MethodSlicer, so we need to take care of that too.
+            elif type(obj) == types.FunctionType:
+                # Check the __qualname__ of the function - if it has
+                # a . then it means it belongs to a class so return
+                # MethodSlicer instead!
+                if hasattr(obj, '__qualname__') and '.' in obj.__qualname__:
+                    log.msg("{} is a method, not a function".format(obj))
+                    slicerFactory = self.slicerTable.get(types.MethodType)
+                else:
+                    slicerFactory = self.slicerTable.get(type(obj))
+                    
+            else:
+                slicerFactory = self.slicerTable.get(type(obj))
+            
         if slicerFactory:
             if self.debug: log.msg(" got slicerFactory %s" % slicerFactory)
             return slicerFactory(obj)
-        if issubclass(type(obj), types.InstanceType):
+
+        if issubclass(type(obj), InstanceType):
             name = str(obj.__class__)
         else:
             name = str(type(obj))
+
         if self.debug: log.msg("cannot serialize %s (%s)" % (obj, name))
         raise Violation("cannot serialize %s (%s)" % (obj, name))
 
     def slice(self):
         return self
+    
     def __iter__(self):
         return self # we are our own iterator
-    def next(self):
+
+    def __next__(self):
         if self.objectSentDeferred:
```

Comment:
wait, don't we alias `__next__` back to `next` above anyway? why the name change?